include/dali/core/geom/vec.h

```diff
@@ -0,0 +1,533 @@
+  using R = decltype(f(vecs[0]...));
+  vec<N, R> result;
+  for (size_t i = 0; i < N; i++) {
+    result[i] = f(vecs[i]...);
+  }
+  return result;
+}
+
+template <typename To, size_t N, typename From>
+DALI_HOST_DEV inline vec<N, To> cast(const vec<N, From> &v) {
+  return v.template cast<To>();
+}
+
+template <size_t N, typename T>
+DALI_HOST_DEV vec<N, T>
+clamp(const vec<N, T> &in, const vec<N, T> &lo, const vec<N, T> &hi) {
+  return elementwise(clamp, in, lo, hi);
+}
+
+template <size_t N, typename T>
+DALI_HOST_DEV vec<N, T> min(const vec<N, T> &a, const vec<N, T> &b) {
+  return elementwise(min, a, b);
+}
+
+template <size_t N, typename T>
+DALI_HOST_DEV vec<N, T> max(const vec<N, T> &a, const vec<N, T> &b) {
+  return elementwise(max, a, b);
+}
+
+#ifdef __CUDA_ARCH__
+template <size_t N>
+__device__ vec<N> floor(const vec<N> &a, const vec<N> &b) {
+  return elementwise(floorf, a);
+}
+
+template <size_t N>
+__device__ vec<N> ceil(const vec<N> &a, const vec<N> &b) {
+  return elementwise(ceilf, a);
+}
+#else
+
+template <size_t N, typename T>
+constexpr vec<N, T> floor(const vec<N, T> &a) {
+  return elementwise(std::floor, a);
+}
+
+template <size_t N, typename T>
+constexpr vec<N, T> ceil(const vec<N, T> &a) {
+  return elementwise(std::ceil, a);
+}
+
+#endif
+
+template <size_t N>
+DALI_HOST_DEV ivec<N> round_int(const vec<N> &a) {
+  return elementwise(static_cast<int(&)(float)>(round_int), a);
+}
+
+template <typename T, size_t size0, size_t size1>
+DALI_HOST_DEV
+constexpr auto cat(const vec<size0, T> &v0, const vec<size1, T> &v1) {
+  vec<size0 + size1, T> ret = {};
+  for (size_t i = 0; i < size0; i ++) {
+    ret[i] = v0[i];
+  }
+  for (size_t i = 0; i < size1; i ++) {
+    ret[i + size0] = v1[i];
+  }
+  return ret;
+}
+
+template <typename T, size_t size0>
+DALI_HOST_DEV
+constexpr auto cat(const vec<size0, T> &v0, T v1) {
+  vec<size0 + 1, T> ret = {};
+  for (size_t i = 0; i < size0; i ++) {
+    ret[i] = v0[i];
+  }
+  ret[size0] = v1;
+  return ret;
+}
+
+template <typename T, size_t size1>
+DALI_HOST_DEV
+constexpr auto cat(T v0, const vec<size1, T> &v1) {
+  vec<size1 + 1, T> ret = {};
+  ret[0] = v0;
+  for (size_t i = 0; i < size1; i ++) {
+    ret[i+1] = v1[i];
+  }
+  return ret;
+}
+
+template <typename T, size_t size0, size_t... sizes>
+DALI_HOST_DEV
+constexpr auto cat(const vec<size0, T> &v0, const vec<sizes, T> &...tail) {
+  return cat(v0, cat(tail...));
+}
+
+template <size_t sub_n, size_t n, typename T>
+DALI_HOST_DEV
+constexpr auto sub(const vec<n, T> &orig, size_t start = 0) {
+  static_assert(sub_n <= n, "Cannot extend a vector using `sub` function.");
+  vec<sub_n, T> ret = {};
+  for (size_t i = 0; i < sub_n; i++)
+    ret[i] = orig[i + start];
+  return ret;
+}
+
+template <typename To, size_t size, typename From,
+          typename OutputType = vec<size*sizeof(From)/sizeof(To), To>>
+DALI_HOST_DEV
+constexpr OutputType bitcast(const vec<size, From> &v) {
```

Comment:
I should add a CI check that searches for bitcast and links to this: https://chromium.googlesource.com/chromium/src/base/+/master/bit_cast.h
:P